tests/run/test_Saver.py

```diff
@@ -0,0 +1,113 @@
+from myrtlespeech.protos import task_config_pb2
+from myrtlespeech.run.callbacks.callback import CallbackHandler
+from myrtlespeech.run.run import Saver
+from myrtlespeech.run.train import fit
+
+from tests.builders.test_task_config import build_and_check_task_config
+from tests.protos.test_task_config import task_configs
+from tests.utils.utils import check_state_dicts_match
+
+# Utilities -------------------------------------------------------------------
+
+
+def change_seq_to_seq_state_dict(seq_to_seq: SeqToSeq):
+    """Alters SeqToSeq state_dict.
+
+    This should ensure that calling this function multiple times should result
+    in a unique state_dict for **every** successive call.
+    """
+    with torch.no_grad():
+        for name, param in seq_to_seq.model.named_parameters():
+            param += 1.0
+    if seq_to_seq.optim:
+        for param_group in seq_to_seq.optim.param_groups:
+            param_group["lr"] += 0.2
+    if seq_to_seq.lr_scheduler:
+        seq_to_seq.lr_scheduler.base_lrs = [
+            x + 0.1 for x in seq_to_seq.lr_scheduler.base_lrs
+        ]
+        seq_to_seq.lr_scheduler._step_count += 2
+
+
+# Tests -----------------------------------------------------------------------
+
+
+@settings(deadline=3000)
+@given(seq_to_seq_cfg=task_configs(), check_with_fit_api=st.booleans())
+def test_seq_to_seq_correctly_built_saved_and_loaded(
+    seq_to_seq_cfg: task_config_pb2.TaskConfig, check_with_fit_api: bool,
+) -> None:
+    """Test that SeqToSeq is correctly built + saved + loaded."""
+    seq_to_seq, epochs, train_loader, _ = build_and_check_task_config(
+        seq_to_seq_cfg
+    )
+
+    with TemporaryDirectory() as tmpdir:
+        # Init Saver Callback, and simulate x2 epochs to save state_dict
+        cb_handler = CallbackHandler(
+            callbacks=[Saver(log_dir=tmpdir, model=seq_to_seq)]
+        )
+
+        cb_handler.on_train_begin(epochs)
+        cb_handler.on_epoch_end()
+        cb_handler.on_epoch_end()  # saves state_dict
+
+        # make copy of saved state_dict
+        expected_sd = copy.deepcopy(seq_to_seq.state_dict())
+
+        # Alter state_dict so that it != expected_sd
+        change_seq_to_seq_state_dict(seq_to_seq)
+
+        # check that state_dict **has** changed by checking that all subdicts
+        # are altered
+        with pytest.raises(AssertionError):
+            check_state_dicts_match(
+                expected_sd["model"], seq_to_seq.state_dict()["model"]
+            )
+        if seq_to_seq.optim:
+            with pytest.raises(AssertionError):
+                check_state_dicts_match(
+                    expected_sd["optim"], seq_to_seq.state_dict()["optim"]
+                )
+        if seq_to_seq.lr_scheduler:
+            with pytest.raises(AssertionError):
+                check_state_dicts_match(
+                    expected_sd["lr_scheduler"],
+                    seq_to_seq.state_dict()["lr_scheduler"],
+                )
```

Comment:
The `check_state_dicts_match` function is recursive so these three calls could be merged into one? The `check_state_dicts_match` function could also return a bool rather than throwing an assertion making it more natural to call?

```python
assert not state_dicts_match(expected_sd, seq_to_seq.state_dict())
```
    
with `check_state_dicts_match` becoming something like (untested):

```python
def state_dicts_match(dict1, dict2) -> bool:
    """Returns True if dicts have same keys and values."""
    if not dict1.keys() == dict2.keys():
        return False
    for key in dict1.keys():
        val1 = dict1[key]
        val2 = dict2[key]
        if isinstance(val1, dict):
            if not (isinstance(val2, dict) and state_dicts_match(val1, val2)):
                return False
        if isinstance(val1, float):
            if not (isinstance(val2, float) and math.isclose(val1, val2)):
                return False
        if isinstance(val1, torch.Tensor):
            if not (isinstance(val2, torch.Tensor) and torch.allclose(val1, val2)):
               return False
    return True
```

Reply:
I've updated it to return a bool but I'm intentionally checking that *all* of the subdicts don't match whereas `assert state_dicts_match(expected_sd, seq_to_seq.state_dict())` will fail if just one of them was changed.